pkg/controller/release/release_controller.go

```diff
@@ -283,34 +272,145 @@ func (c *Controller) syncOneReleaseHandler(key string) error {
+	rel, err := c.scheduleRelease(initialRel)
+	if err != nil {
+		return err
+	}
 
-	if !reflect.DeepEqual(initialRel, rel) {
-		if _, err := c.clientset.ShipperV1alpha1().Releases(namespace).Update(rel); err != nil {
-			return shippererrors.NewKubeclientUpdateError(rel, err).
-				WithShipperKind("Release")
+	if !equality.Semantic.DeepEqual(initialRel, rel) {
+		var updErr error
+		rel, updErr = c.clientset.ShipperV1alpha1().Releases(rel.Namespace).Update(rel)
+		if updErr != nil {
+			return shippererrors.NewKubeclientUpdateError(rel, updErr).WithShipperKind("Release")
 		}
 	}
 
+	patches, trans, err := c.ensureReleaseState(rel.DeepCopy())
 	if err != nil {
 		return err
 	}
 
-	appKey, err := c.getAssociatedApplicationKey(rel)
-	if err != nil {
-		return err
+	for _, t := range trans {
+		c.recorder.Eventf(
+			rel,
+			corev1.EventTypeNormal,
+			"ReleaseStateTransitioned",
+			"Release %q had its state %q transitioned to %q",
+			shippercontroller.MetaKey(rel),
+			t.State,
+			t.New,
+		)
 	}
 
-	// If everything went fine, scheduling an application key in the
-	// application workqueue.
-	klog.V(4).Infof("Scheduling Application key %q", appKey)
-	c.applicationWorkqueue.Add(appKey)
+	for _, patch := range patches {
+		if err := c.applyPatch(namespace, patch); err != nil {
+			return err
+		}
+	}
 
 	klog.V(4).Infof("Done processing Release %q", key)
 
 	return nil
 }
 
+func (c *Controller) applicationReleases(rel *shipper.Release) ([]*shipper.Release, error) {
+	appName, err := releaseutil.ApplicationNameForRelease(rel)
+	if err != nil {
+		return nil, err
+	}
+	releases, err := c.releaseLister.Releases(rel.Namespace).ReleasesForApplication(appName)
+	if err != nil {
+		return nil, err
+	}
+	return releases, nil
+}
+
+func (c *Controller) ensureReleaseState(curr *shipper.Release) ([]ExecutorResult, []ReleaseStrategyStateTransition, error) {
+	releases, err := c.applicationReleases(curr)
+	if err != nil {
+		return nil, nil, err
+	}
+	prev, succ, err := releaseutil.GetSiblingReleases(curr, releases)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	relInfoCurr, err := c.buildReleaseInfo(curr)
+	if err != nil {
+		return nil, nil, err
+	}
+	var relInfoPrev, relInfoSucc *releaseInfo
+	if prev != nil {
+		relInfoPrev, err = c.buildReleaseInfo(prev)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	if succ != nil {
+		relInfoSucc, err = c.buildReleaseInfo(succ)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	//TODO: this flag is deprecated, we keep it here for the sake of gradual
+	//changes. The flag propagates as deep as strategy conditions
+	//consolidation functions
+	// see pkg/util/conditions/strategy.go for more details.
+	hasIncumbent := len(releases) > 1
+	executor := NewStrategyExecutor(relInfoCurr, relInfoPrev, relInfoSucc, c.recorder, hasIncumbent)
+
+	patches, trans, err := executor.Execute()
+	if err != nil {
+		releaseSyncedCond := releaseutil.NewReleaseCondition(
+			shipper.ReleaseConditionTypeScheduled,
```

Comment:
Maybe `Scheduled` isn't the right condition here. When this error triggers, shipper will enter into a loop as the scheduler will set this back to `True`, this will change it to `False`, trigger a new sync, repeat.